### PR TITLE
001 view job roles

### PR DESCRIPTION
**Pull request by cderry1**

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -23,6 +23,6 @@ module.exports = {
         "project": ["tsconfig.json"]
     },
     "rules": {
-        
+        "@typescript-eslint/no-misused-promises": 1
```
Comment (david-andrew-k) [inline .eslintrc.js:26]: Why was this added back?

File: controllers/jobRoleController.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  } catch (e) {
+    data = []
+    console.error(e)
+  }
+
+  res.render('job-roles', {
+    roles: data
+  })
+}
+/*
```
Comment (david-andrew-k) [inline controllers/jobRoleController.ts:32]: Remove this commented code

File: routes/routes.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+/* eslint-disable @typescript-eslint/no-misused-promises */
+
+import express, { type Router } from 'express'
+import { getJobRolesResponse, getJobRoleResponse } from '../controllers/jobRoleController'
+
+const router: Router = express.Router()
+
+router.get('/job-roles', getJobRolesResponse)
+router.get('/job-role/:id', getJobRoleResponse)
```
Comment (david-andrew-k) [inline routes/routes.ts:9]: Remove ID route

File: test/unit/service/jobRoleServiceTest.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        var mock = new MockAdapter(axios);
+
+        mock.onGet(JobRoleService.URL).reply(500);
+
+        var error = await JobRoleService.getJobRoles()
+        console.log(error.message)
+        expect(error.message).to.equal('Could not get Job Roles')
+      })
+    })
+   /* describe('getJobRoleById', function () {
```
Comment (david-andrew-k) [inline test/unit/service/jobRoleServiceTest.js:38]: Remove this comment

File: views/view-job-role.html [hunk 5]
```diff
@@ -0,0 +1,43 @@
+<!DOCTYPE html>
```
Comment (david-andrew-k) [inline views/view-job-role.html:1]: Remove this file - Not relevant to user story

**Review by david-andrew-k — APPROVED** [review]
Looks Good Conor!

File: app.ts [hunk 6]
```diff
@@ -36,3 +36,7 @@ app.get('/', (req: Request, res: Response) => {
 
   })
 })
+
+// jobRole(app)
+
```
Comment (ccurran01) [inline app.ts:41]: Is this required?

File: controllers/jobRoleController.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
+import { type Request, type Response } from 'express'
+import { getJobRoles } from '../src/services/jobRoleService'
+import { type JobRole } from '../models/jobRole'
+
+export const getJobRolesResponse = async (req: Request, res: Response): Promise<void> => {
+  let data: JobRole[]
+
+  try {
+    data = await getJobRoles()
+    console.log(data)
```
Comment (ccurran01) [inline controllers/jobRoleController.ts:10]: Should this be getting logged?

File: package.json [hunk 8]
```diff
@@ -10,9 +10,10 @@
     "lint": "npx eslint . --ext .ts && echo \"\n### Linting Successful - No standards violations detected! ###\n\"",
     "lintFix": "npx eslint . --ext .ts --fix",
     "test": "echo \"Error: no test specified yet\" && exit 0",
+    "testIntegration": "mocha test/unit/**/*.js",
```
Comment (ccurran01) [inline package.json:13]: the command is called testIntegration but it looks like it kicks off the unit tests?

File: package.json [hunk 9]
```diff
@@ -10,9 +10,10 @@
     "lint": "npx eslint . --ext .ts && echo \"\n### Linting Successful - No standards violations detected! ###\n\"",
     "lintFix": "npx eslint . --ext .ts --fix",
     "test": "echo \"Error: no test specified yet\" && exit 0",
```
Comment (ccurran01) [inline package.json:12]: Is this required?

File: package.json [hunk 10]
```diff
@@ -35,10 +36,17 @@
...
+    "axios-mock-adapter": "^1.21.4",
+    "chai": "^4.3.7",
+    "mocha": "^10.2.0",
+    "selenium-webdriver": "^4.9.0"
   },
   "dependencies": {
+    "axios": "^1.4.0",
     "express": "^4.18.2",
-    "nunjucks": "^3.2.4"
+    "mocha": "^10.2.0",
```
Comment (ccurran01) [inline package.json:48]: should mocha be declared as a dependency?

File: src/services/jobRoleService.ts [hunk 11]
```diff
@@ -0,0 +1,12 @@
+import axios from 'axios'
+import { type JobRole } from '../../models/jobRole'
+
+export const getJobRoles = async function (): Promise<JobRole[]> {
+  try {
+    // Add API Endpoint
+    const response = await axios.get('http://localhost:8080/api/job-roles')
```
Comment (ccurran01) [inline src/services/jobRoleService.ts:7]: Could we extract the `http://localhost:8080` base url out to be an environment variable? This should make things easier when we come to get the app deployed to AWS

File: src/services/jobRoleService.ts [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+    return response.data
+  } catch (e) {
+    throw new Error('Could not get Job Roles')
```
Comment (ccurran01) [inline src/services/jobRoleService.ts:10]: Could this be a custom error?

File: models/jobRole.ts [hunk 13]
```diff
@@ -0,0 +1,7 @@
+export interface JobRole {
+  id?: number
+  name?: string
+  specificationDescription?: string
+  bandName?: string
+  capability?: string
+}
```
Comment (ccurran01) [inline models/jobRole.ts:7]: Are all these properties optional?

File: test/unit/service/jobRoleServiceTest.js [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    name: "Engineer",
+    specificationDescription: "Q/A testing and Software",
+    capability: "Amazing",
+    BandName: "Amazing",
+}
+
+describe('JobRoleService', function () {
+    describe('getJobRoles', function () {
+      it('should return Job Roles from response', async () => {
+        var mock = new MockAdapter(axios);
```
Comment (ccurran01) [inline test/unit/service/jobRoleServiceTest.js:17]: var keyword doesn't really get used any more. Could we update the use of vars to let or const, please

This might be a good rule to add to our lint config. Think we could give this one a go - https://eslint.org/docs/latest/rules/no-var

File: test/unit/service/jobRoleServiceTest.js [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+        expect(results[0]).to.deep.equal(job)
+      })
+
+      it('should throw exception when 500 error returned from axios', async () => {
+        var mock = new MockAdapter(axios);
+
+        mock.onGet(JobRoleService.URL).reply(500);
+
+        var error = await JobRoleService.getJobRoles()
+        console.log(error.message)
```
Comment (ccurran01) [inline test/unit/service/jobRoleServiceTest.js:34]: Is this required?

File: views/job-roles.html
Comment (ccurran01) [inline views/job-roles.html:1]: Have you scanned the page with Wave to check for accessibility issues?

File: test/unit/service/jobRoleServiceTest.js [hunk 16]
```diff
@@ -0,0 +1,38 @@
+var axios = require('axios');
+var MockAdapter = require('axios-mock-adapter');
+var chai = require('chai');  
+const expect = chai.expect;
+const JobRoleService = require('../../../dist/src/services/jobRoleService');
+const job = {
```
Comment (ccurran01) [inline test/unit/service/jobRoleServiceTest.js:6]: What you think about having a line break here?

```suggestion
const JobRoleService = require('../../../dist/src/services/jobRoleService');

const job = {
```

Comment (ccurran01) [inline test/unit/service/jobRoleServiceTest.js:6]: Should this have a type?

File: controllers/jobRoleController.ts [hunk 17]
```diff
@@ -0,0 +1,19 @@
+import { type Request, type Response } from 'express'
+import { getJobRoles } from '../src/services/jobRoleService'
+import { type JobRole } from '../models/jobRole'
+
+export const getJobRolesResponse = async (req: Request, res: Response): Promise<void> => {
```
Comment (ccurran01) [inline controllers/jobRoleController.ts:5]: There doesn't seem to be any test coverage for this function

File: test/unit/service/jobRoleServiceTest.js (hunk 16 above)
Comment (cderry1) [reply test/unit/service/jobRoleServiceTest.js]: I didn't know how to assign it a type

File: controllers/jobRoleController.ts (hunk 17 above)
Comment (cderry1) [reply controllers/jobRoleController.ts:5]: I don't exactly know how you would test the Controller method.

File: test/unit/service/jobRoleServiceTest.js (hunk 16 above)
Comment (ccurran01) [reply test/unit/service/jobRoleServiceTest.js]: It looks to me like this is a job role object to help us out in tests? In typescript, the type comes after the variable name like this

```suggestion
const job: JobRole = {
```

File: test/unit/jobRoleControllerTest.ts [hunk 18]
```diff
@@ -0,0 +1,45 @@
+import { expect } from 'chai'
+import { getJobRolesResponse } from '../../src/controllers/jobRoleController'
+interface JobRole {
+  id: 1
+  name: 'Engineer'
+  specificationDescription: 'Q/A testing and Software'
+  capability: 'Amazing'
+  bandName: 'Amazing'
+}
```
Comment (ccurran01) [inline test/unit/jobRoleControllerTest.ts:9]: Is this required?

**Review by ccurran01 — COMMENTED** [review]
Looking good @david-andrew-k, @cderry1. Doesn't look like that interface is doing much. If we get that tidied up it's good to go 👍

**Review by ccurran01 — APPROVED** [review]
